autopilot/scanner.go

```diff
@@ -2,89 +2,278 @@ package autopilot
+		}
+
+		pool    *scanPool
+		tracker *tracker
+
+		scanMinInterval    time.Duration
+		timeoutMinInterval time.Duration
+
+		stopChan chan struct{}
+
+		mu                sync.Mutex
+		scanning          bool
+		scanningLastStart time.Time
+		timeout           time.Duration
+		timeoutLastUpdate time.Time
+	}
+
+	scanReq struct {
+		hostKey consensus.PublicKey
+		hostIP  string
+	}
+
+	scanResp struct {
+		hostKey  consensus.PublicKey
+		settings rhpv2.HostSettings
+		err      error
+	}
+
+	scanPool struct {
+		s          *scanner
+		numThreads uint8
+		scanQueue  chan scanReq
+	}
+
+	tracker struct {
+		threshold  uint64
+		percentile float64
 
-	mu      sync.Mutex
-	running bool
+		mu      sync.Mutex
+		cnt     uint64
+		timings []float64
+	}
+)
+
+func newScanPool(s *scanner, threads uint8) *scanPool {
+	return &scanPool{numThreads: threads, s: s}
+}
+
+func (p *scanPool) launchScans(hosts []hostdb.Host) {
+	p.scanQueue = make(chan scanReq, p.numThreads)
+	go func() {
+		defer close(p.scanQueue)
+		for _, h := range hosts {
+			if p.s.isStopped() {
+				return
+			}
+			p.scanQueue <- scanReq{
+				hostKey: h.PublicKey,
+				hostIP:  h.NetAddress(),
+			}
+		}
+	}()
 }
 
-func newScanner(ap *Autopilot) *scanner {
-	return &scanner{
-		ap: ap,
+func (p *scanPool) launchWorkers(workerFn func(chan scanResp)) chan scanResp {
+	respChan := make(chan scanResp, p.numThreads)
+	for i := uint8(0); i < p.numThreads; i++ {
+		go workerFn(respChan)
 	}
+	return respChan
 }
 
-func (s *scanner) tryPerformHostScan() {
+func newTracker(min, total uint64, percentile float64) *tracker {
+	return &tracker{
+		threshold:  min,
+		percentile: percentile,
+		timings:    make([]float64, total),
+	}
+}
+
+func (t *tracker) addTiming(timing float64) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.timings[t.cnt%uint64(len(t.timings))] = timing
+	if t.cnt == math.MaxUint64 {
+		t.cnt = 0
+	} else {
+		t.cnt += 1
+	}
+}
+
+func (t *tracker) timeout() (time.Duration, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if t.cnt < uint64(t.threshold) {
+		return 0, nil
+	}
+
+	percentile, err := stats.Percentile(t.timings, t.percentile)
+	if err != nil {
+		return 0, err
+	}
+
+	return time.Duration(percentile) * time.Millisecond, nil
+}
+
+func newScanner(ap *Autopilot, threads uint8, scanMinInterval, timeoutMinInterval time.Duration) *scanner {
+	s := &scanner{
+		bus:    ap.bus,
+		worker: ap.worker,
+
+		tracker: newTracker(trackerMinDataPoints, trackerNumDataPoints, trackerTimeoutPercentile),
+
+		stopChan: ap.stopChan,
+
+		scanMinInterval:    scanMinInterval,
+		timeoutMinInterval: timeoutMinInterval,
+	}
+	s.pool = newScanPool(s, threads)
+	return s
+}
+
+func (s *scanner) tryPerformHostScan() bool {
 	s.mu.Lock()
-	if s.running {
+	if s.scanning || !s.isScanRequired() {
 		s.mu.Unlock()
-		return
+		return false
 	}
-	s.running = true
+
+	cs, _ := s.bus.ConsensusState()
+	if !cs.Synced {
+		s.mu.Unlock()
+		return false
```

Comment:
can't we perform host scans without being synced?

Reply:
Yeah this is mistakenly carried over from `skyd` where there's the concept of an "initial scan". The idea was to only perform the initial scan when we're synced to ensure all hosts are part of that initial scan.